### DISCUSSION TITLE
Capture Image with External Lighting

### DISCUSSION THREAD

**Discussion by waikeat2397** [raspberrypi/picamera2#1043] [Q&A] [answered]
I have limited resources, thus I am doing something not a very good practice, but still find it weird. 

I am using API to control a microcontroller to turn on external lighting. 

I wait 50ms before request a frame from camera. 

Sample code as below.

```
from lighting_mcu import *
from libcamera import controls
from picamera2 import Picamera2

import time

mcu = LightingController()

picam2 = Picamera2()
config = picam2.create_still_configuration(main={"size": (4056, 3040), 'format': 'BGR888'}, raw={"size": (4056, 3040)})
picam2.configure(config)

picam2.start()
picam2.set_controls({'AeEnable': False})
picam2.set_controls({'AwbEnable': False})
picam2.set_controls({'ExposureTime': 900})
picam2.set_controls({'ColourGains': (3.047791004180908, 1.556588888168335)})
time.sleep(2)


x = 0
while True:
    mcu.open_light()
    time.sleep(0.05)
    picam2.capture_file(f'save.jpg')
    mcu.close_light()
    print(f"Cycle {x}")
    x = x + 1

picam2.stop()
picam2.close()
```

FYI, i am using HQ camera. Everything ok when I using resolution (2028, 1520), the effect is good.

But when I use full resolution (4056, 3040), the image turn out part of it is dark. 

I think it should have same effect as (2028, 1520) resolution as both mode is using full sensor pixel with same crop limit. 

Is ti because of exposure? But I have already turn off auto exposure and set a fixed exposure value.

Below are the sample image.

![13](https://github.com/raspberrypi/picamera2/assets/122470067/11db70ce-25f2-4494-b1da-7ebaca06cc47)

**Comment by waikeat2397** [comment]
Ahhh I think i got the idea already. Seems like I got the frame when the lighting is not fully open. 
The time taken for my lighting to fully open is at around 30ms.

Under full resolution, HQ camera should able to run in 5fps, which means around 200ms per frame.

![image](https://github.com/raspberrypi/picamera2/assets/122470067/0a6e0887-b4cd-4074-991c-1dcd718a1855)

Based on documentation, i should get one new frame upon requests instead getting and old frame.
In this case why am I getting frame where lighting is not fully open which is an old frame? Anything I missed out here?

**Comment by davidplowman** [comment, marked as answer]
In general, the camera system is quite heavily pipelined, so the time you receive a frame doesn't tell you much about when it was captured. There is actually a `"SensorTimestamp"` field in the image metadata that tells you when the first pixel of the frame was received.

But even that isn't good enough, because the image will start to be exposured _well before_ the first pixel is read out. The Picamera2 `capture_request()` function actually has a feature to handle this. If you pass the `flush` parameter as `True`, then it will guarantee the no pixel is exposed before that moment. So your code would look something like
```
    request = picam2.capture_request(flush=True)
    request.save('main', "save.jpg")
    request.release()
```
